### PR TITLE
SSE-3210: Jira integration for Register Your Interest Form

### PR DESCRIPTION
**Pull request by Ryan-Andrews99**
Dependent on: https://github.com/govuk-one-login/onboarding-self-service-infra/pull/21

## Why

Enables an integration requested by the onboarding team.

Currently we take data from the register your interest form and send it to a Google Sheet for staff to work on.
We are looking to automate aspects of that process by sending the data directly into 
Jira instead.

## What

[Screencast from 12-03-24 14:22:31.webm](https://github.com/govuk-one-login/onboarding-product-page/assets/3694062/6812b956-ea94-4697-af0f-5e4a7bbe1d9e)

This PR:
- Adds a new Jira service to handle posting to create issues
- Creates a stub to enable integration testing
- Masks the real and stub interface behind a single class, sets the toggle between functions as an Env Var
- Adds the integration to the register your interest control with a feature flag
- Adds a feature flag defaulted to ON for the google sheets feature

It should look like this:

## How to Test
- Whilst logged into Jira [visit the Atlassian Dashboard](https://id.atlassian.com/manage-profile/security) and register yourself a personal API token.
- Copy across the new .env.example into your .env and populate with your work email as username and the token you just created as the credential
- Populate the env with JIRA_PROJECT_KEY=APIT and JIRA_ISSUE_TYPE=10907)
- Spin up the local application and processed to fill out the reigster your interest form
- Submit and check a value as appeared on the [test board](https://govukverify.atlassian.net/jira/software/projects/APIT/boards/510)
- Turn the feature flag off in the ENV for Jira, re-test and check no calls / board values appear
- Turn off the feature flag for google sheets and check from the logs that the stub has not been called

After that review code normally

### CODE REVIEW

**Comment by huwd** [comment]
@Ryan-Andrews99 I think this is ready for the next step.

Just before we get a review let's see if we can sort out the deployment + sonar cloud issues. Will pick up tomorrow.

File: infrastructure/frontend/frontend.template.yml [hunk 1]
```diff
@@ -142,6 +142,8 @@ Resources:
               Value: !Ref ContainerPort
             - Name: SHOW_TEST_BANNER
               Value: !Ref ShowTestBanner
+            - Name: GOOGLE_SHEETS_INTEGRATION_ENABLED
```
Comment (huwd) [inline infrastructure/frontend/frontend.template.yml:145]: @kalpaitch Wonder if I can get a sense check from you on this commit?

Context: The change we're adding involves some new  ENV Vars.
They fall into a few categories:
- x2 feature flags (should differ between environments but are not secret)
- x1 new stub toggle (similar story, but also same as what we do with sheets + service now)
- x3 configuration variables (I've hard coded these, they should be the same everywhere, there's no integration Jira for now, but flexible for future use)
- x2 secrets (these could vary by env, but are secret and so I think we need them in SSM)

I wouldn't mind a sense check that I'm using cloudformation correctly here.
I've also not created any SSM params in any env yet.

Would appreciate a quick chat about how best to do that, avoiding clickops wherever possible for the audit trail?

**Review by huwd — COMMENTED** [review]
@kalpaitch / @rafal-surowiec-co 
I'm still a little bit fuzzy on our branch deploy process. I suspect me changing the cloudformation templates has caused issues there...

Any thoughts on what I should check to get those deploys working?

**Review by Ryan-Andrews99 — COMMENTED** [review]
1 Minor comment just around making some of the error logging a bit clearer. Obviously I can't approve 😆,  this but ✅ from me

File: infrastructure/frontend/frontend.template.yml (hunk 1 above)
Comment (rafal-surowiec-co) [reply infrastructure/frontend/frontend.template.yml:145]: Hardcoded values could be ssm param. This is a good practice.

File: src/controllers/register.ts [hunk 2]
```diff
@@ -113,16 +115,35 @@ export const post = async function (req: Request, res: Response) {
...
+            try {
+                const jiraService = new JiraTicketService(values);
+                await jiraService.sendJiraTicket();
+            } catch (error) {
+                if (axios.isAxiosError(error)) {
+                    console.error("Jira Post request failed with status code: ", error.status);
+                    console.error("Axios Error, errorMessages Object: ", error.response?.data?.errorMessages);
+                    console.error("Axios Error, errors Object: ", error.response?.data?.errors);
+                } else {
+                    console.error("Jira Integration error: ", error);
```
Comment (rafal-surowiec-co) [inline src/controllers/register.ts:142]: When logging errors please make sure we are not logging PII data.

Comment (rafal-surowiec-co) [reply src/controllers/register.ts:142]: Please squash commits

Comment (huwd) [reply src/controllers/register.ts:142]: @Ryan-Andrews99 you happy to address this one?

**Comment by huwd** [comment]
We've also got a linting and deploy issue.
@Ryan-Andrews99 would you be free to follow those up too?

File: src/controllers/register.ts (hunk 2 above)
Comment (Ryan-Andrews99) [reply src/controllers/register.ts:142]: I think the error logging here strikes the right balance. We're only logging parts of the Axios error which will not log the POST payload which all the PII. The `errors` and `errorMessages` objects _may_ contain PII, depending on the Jira API response but they also will likely contain very useful information about why a request might've been rejected. Given we've caught the Axios error in the `if` the only other errors we're likely to log are missing environment variable names, which do not contain PII.

**Comment by huwd** [comment]
Excuse the fairly brutal last force push there,
@Ryan-Andrews99 and I had a chat.

I think it's right to squash the last commit into the main register controller commit (i've done that).
But I'm a big believer on not squashing an entire PR down into a single commit.

If we've done our commits right they should [be atomic](https://gds-way.digital.cabinet-office.gov.uk/standards/source-code/working-with-git.html#atomic-commits) and independently revertible.

 This should help with aiding legacy understanding of our repo + maintainability and incident response

**Comment by huwd** [comment]
Linting issue relates to #333

File: infrastructure/frontend/frontend.template.yml (hunk 1 above)
Comment (huwd) [reply infrastructure/frontend/frontend.template.yml:143]: That's a good set of rules of thumb, thanks @kalpaitch and @rafal-surowiec-co

Comment (Ryan-Andrews99) [reply infrastructure/frontend/frontend.template.yml:143]: Have moved the static values into a config file `config/jiraConfig.ts`, so I'm gonna mark this one as resolved unless there are any objections

File: src/config/jiraConfig.ts [hunk 3]
```diff
@@ -0,0 +1,3 @@
+export const JIRA_BOARD_URL = "https://govukverify.atlassian.net/rest/api/3/issue";
```
Comment (rafal-surowiec-co) [inline src/config/jiraConfig.ts:1]: SSM Params ?

Comment (huwd) [reply src/config/jiraConfig.ts:1]: What's the advantage of SSM params here over hard coded?

These don't vary be environment, which would be our usual benefit.
They're not secret.
And to change them in config would mean having to commit code, merge and release code elsewhere, so there's a similar go-to-live cadence.

To me, makes more sense to have these next to the code that uses them, with good commits and a version control history. Especially as to run them locally these are values that turn up in .env and someone might want to snoop the integration to work out how.

I figure the lower complexity option is good value here?

Comment (rafal-surowiec-co) [reply src/config/jiraConfig.ts:1]: Changing SSM param does not require code change or commit. In our case it is updating value manually in AWS console and restarting pipeline.

Comment (huwd) [reply src/config/jiraConfig.ts:1]: Do we not populate non-secret SSM params via Infrastructure as Code?
If not I would have thought the version control history is the real asset here?

We should try and avoid clickops where we can?

**Comment by huwd** [comment]
> Please squash commits, ideally into two groups: [SSE-3210](https://govukverify.atlassian.net/browse/SSE-3210) and AP-191.

I don't think squashing commits here is the right approach.
I think this history and these messages are really valuable to be in our `main` branch.

Both to preserve the long term intent behind our code and it's changes, but also to offer fairly granular rollback points.
This is part of what we're after when the GDS way [advises atomic commits](https://gds-way.digital.cabinet-office.gov.uk/standards/source-code/working-with-git.html#commits).

It's also something we've been doing work on recently @BeckaL and I also [gave an internal talk about this recently.](https://docs.google.com/presentation/d/1eQYnrdGBiE9DskR4UDhJRp4NSTgwHV7Ah3-LuC4Cq6Y/edit#slide=id.g10d42026b8_2_0)
So there's a drive right now across the program to get more of this raw and unsquashed history into our main branches.

Hope that makes sense?
Happy to pick up and chat if you disagree?

**Review by rafal-surowiec-co — APPROVED** [review]

**Review by huwd — APPROVED** [review]

**Pull request merged**